jctools-core/src/main/java/org/jctools/queues/MpscBlockingConsumerArrayQueue.java

```diff
@@ -521,7 +554,7 @@ public E poll(long timeout, TimeUnit unit) throws InterruptedException
         return e;
     }
 
-    private E parkUntilNext(E[] buffer, long cIndex, long offset, long timeoutNs) throws InterruptedException {
+    private E parkUntilNext(E[] buffer, long cIndex, long offset, long timeoutNs, boolean peek) throws InterruptedException {
```

Comment:
I'm not a fan of adding further complexity to `parkUntilNext`.

I was previously considering splitting out a version without the timeout for `take`; and removing the ternary checks for timeoutNs = Long.MAX_VALUE. (The downside to this is that you might have an unnecessary call  to System.nanoTime() if the poll timeout is greater than Long.MAX_VALUE ns ~ 300 years, but the caller can always use `poll()`).

If we did this, then we could also separate out and inline the appropriate consumption behaviour in the tail without the `peek` boolean.

Reply:
Given that's a supposed to be a slowpath method I didn't care much about adding a single check there, but I am more concerned about the added API; if it's welcome and make sense and if it can be useful to users. Any feedback re this?

Thanks for the review 😊

Reply:
I agree with the consensus on blockingDrain.